sdks/java/core/src/main/java/org/apache/beam/sdk/util/GcsUtil.java

```diff
@@ -372,10 +402,11 @@ public boolean shouldRetry(IOException e) {
             },
             IOException.class,
             sleeper);
-        return true;
+
+        return bucket;
       } catch (GoogleJsonResponseException e) {
         if (errorExtractor.itemNotFound(e) || errorExtractor.accessDenied(e)) {
-          return false;
+          return null;
```

Comment:
This seems like it should be wrapped in a more specific exception and rethrown, given how we're using it